src/ApiHandler.ts

```diff
@@ -124,20 +130,81 @@ export default class ApiHandler {
 			}
 		}
 
+		const perBill = new BN(1e9);
+		const perByteFee = api.consts.transactionPayment.transactionByteFee.mul(perBill);
+		const extrinsicBaseWeight = api.consts.system.extrinsicBaseWeight.mul(perBill);
+		const coefficients = api.consts.transactionPayment.weightToFee[0];
+		const coeffFrac = coefficients.coeffFrac.mul(perBill);
+		const coeffInt = coefficients.coeffInteger.mul(perBill);
+
+		// bn.js values must be < 0x20000000000000, cannot divide by 1e18 [Ref: bn.js:128]
+		const multiplier = (await api.query.transactionPayment.nextFeeMultiplier.at(parentHash)).div(perBill).div(perBill);
+		
+		// TODO: We assume that the polynomial has one positive term of degree one.
+		// const degree = coefficients.degree;
+		// const negative = coefficients.negative;
+
 		for (let idx = 0; idx < block.extrinsics.length; ++idx) {
 			if (!extrinsics[idx].paysFee || !block.extrinsics[idx].isSigned) {
 				continue;
 			}
 
 			try {
-				// This is only a temporary solution. This runtime RPC will not work if types or logic
-				// involved in fee calculation changes in a runtime upgrade. For a long-term solution,
-				// we need to calculate fees based on the metadata constants and fee multiplier.
-				// https://github.com/paritytech/substrate-api-sidecar/issues/45
-				extrinsics[idx].info = api.createType(
-					'RuntimeDispatchInfo',
-					await api.rpc.payment.queryInfo(block.extrinsics[idx].toHex(), parentHash)
-				);
+				if (api.runtimeVersion.specName.toString() === 'kusama') {
+					extrinsics[idx].info = api.createType(
+						'RuntimeDispatchInfo',
+						await api.rpc.payment.queryInfo(block.extrinsics[idx].toHex(), parentHash)
```

Comment:
ah my bad, there was a small misunderstanding.

`api.rpc.payment.queryInfo()` will return a RuntimeDispatchInfo JS class, so there's no need to api.createType here, it's a no-op.

(continued below)